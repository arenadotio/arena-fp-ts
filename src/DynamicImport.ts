/**
 * @since 0.0.1
 */

import * as O from 'fp-ts/lib/Option';
import * as TO from 'fp-ts/lib/TaskOption';
import * as S from 'fp-ts/lib/Semigroup';
import * as M from 'fp-ts/lib/Monoid';
import { pipe } from 'fp-ts/lib/function';

// -------------------------------------------------------------------------------------
// model
// -------------------------------------------------------------------------------------

/**
 * @category model
 * @since 0.0.1
 */
export type DynamicImport<T = any> = TO.TaskOption<T>;

// -------------------------------------------------------------------------------------
// constructors
// -------------------------------------------------------------------------------------

/**
 * @category constructors
 * @since 0.0.1
 */

export function fromModule<A = any>(name: string): TO.TaskOption<A> {
    return async () => {
        try {
            return O.some(await import(name));
        } catch {
            return O.none;
        }
    }
}

// -------------------------------------------------------------------------------------
// instances
// -------------------------------------------------------------------------------------

/**
 * @internal
 */
const _concat = <A>(first: DynamicImport<A>, second: DynamicImport<A>): TO.TaskOption<A> => pipe(
    first,
    TO.alt(() => second),
);

/**
 * @category instances
 * @since 0.0.1
 */
export const Semigroup: S.Semigroup<DynamicImport> = {
    concat: _concat,
};

/**
 * @category instances
 * @since 0.0.1
 */
export const Monoid: M.Monoid<DynamicImport> = {
    concat: Semigroup.concat,
    empty: TO.none,
};

// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------

/**
 * @category utils
 * @since 0.0.1
 */
export function firstSome<T = any>(...mas: ReadonlyArray<DynamicImport<T>>): DynamicImport<T> {
    return M.concatAll(Monoid)(mas);
}
