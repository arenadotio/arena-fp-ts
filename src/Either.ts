/**
 * @since 0.0.10
 */

import * as E from 'fp-ts/lib/Either';
import { flow, LazyArg } from 'fp-ts/lib/function';
import { Validation } from 'io-ts';
import { formatValidationErrors } from 'io-ts-reporters';
import * as Str from 'fp-ts-std/String';

// -------------------------------------------------------------------------------------
// models
// -------------------------------------------------------------------------------------

/**
 * @category models
 * @since 0.0.10
 */
export type Either<A> = E.Either<Error, A>;

// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------

/**
 * @category utils
 * @since 0.0.10
 */
export const fromValidation: <A>(ma: Validation<A>) => Either<A> = flow(
  E.mapLeft(formatValidationErrors),
  E.mapLeft(Str.unlines),
  E.mapLeft(Str.prepend('Validation Failed\n')),
  E.mapLeft((str) => new Error(str))
);

/**
 * @category utils
 * @since 0.0.10
 */
export const tryCatch = <A>(f: LazyArg<A>) => E.tryCatch(f, E.toError);
