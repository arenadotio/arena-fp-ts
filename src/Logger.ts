/**
 * `Logger` is a type re-exported from `pino` and can be used to wrap calls to
 * `pino` in `IO<void>` instances. Using the `State` monad, logging functions
 * can be composed while preserving changes they make to the logger bindings.
 *
 * @example
 * import pino from 'pino';
 *
 * import { pipe } from 'fp-ts/lib/function';
 * import * as IO from 'fp-ts/lib/IO';
 * import * as S from 'fp-ts/lib/State';
 * import * as A from 'fp-ts/lib/Apply';
 *
 * import * as L from '@arenadotio/arena-fp-ts/Logger';
 *
 * // If you have some general business logic functions
 * const f = (s: string) => s.toUpperCase();
 * const g = (n: number) => n * n;
 * const h = (b: boolean) => !b;
 *
 * // Create functions which produce State monads that can perform the logging
 * const step1 = (
 *     s: string,
 * ): S.State<L.Logger, IO.IO<string>> =>(
 *     logger,
 * ) => {
 *     const log = L.info(logger);
 *     // Return a tuple of the result and the next logger
 *     return [
 *         // Result logs a message, then returns the value of f(s)
 *         pipe(
 *             log('In step one with arg %s', s),
 *             IO.apSecond(() => f(s)),
 *         ),
 *         // Create a new logger for step2
 *         logger.child({ step1: 'called' })
 *     ]
 * }
 *
 * const step2 = (
 *     n: number,
 * ): S.State<L.Logger, IO.IO<number>> =>(
 *     logger,
 * ) => {
 *     const log = L.debug(logger);
 *     // Return a tuple of the result and the next logger
 *     return [
 *         // Result logs a message, then returns the value of f(s)
 *         pipe(
 *             log('In step two with arg %d', n),
 *             IO.apSecond(() => g(n)),
 *         ),
 *         // Create a new logger for step2
 *         logger.child({ step2: 'called' })
 *     ]
 * }
 *
 * const step3 = (
 *     b: boolean,
 * ): S.State<L.Logger, IO.IO<boolean>> =>(
 *     logger,
 * ) => {
 *     const log = L.info(logger);
 *     // Return a tuple of the result and the next logger
 *     return [
 *         // Result logs a message, then returns the value of h(b)
 *         pipe(
 *             log('In step three with arg %s', String(b)),
 *             IO.apSecond(() => h(b)),
 *         ),
 *         // Create a new logger for step2
 *         logger.child({ step3: 'called' })
 *     ]
 * }
 *
 * // Sequence the State monads
 * const [ios, finalLogger] = pipe(
 *     pino(),
 *     S.sequenceArray<L.Logger, IO.IO<string | number | boolean>>([
 *         step1('foo'),
 *         step2(2),
 *         step3(true)
 *     ]),
 * );
 *
 * // Sequence the IOs into a single function
 * const program = pipe(
 *     // Convert the array into a tuple
 *     [ios[0]!, ...ios.slice(1)] as const,
 *     (args) => A.sequenceT(IO.Apply)(...args),
 * )
 *
 * assert.deepStrictEqual(program(), ['FOO', 4, false]);
 * assert.deepStrictEqual(
 *     finalLogger.bindings(),
 *     {
 *         step1: 'called',
 *         step2: 'called',
 *         step3: 'called',
 *     }
 * )
 *
 * @since 0.0.1
 */

import Pino from 'pino';
import { IO } from 'fp-ts/lib/IO';

// -------------------------------------------------------------------------------------
// model
// -------------------------------------------------------------------------------------

/**
 * @category model
 * @since 0.0.1
 */
export type PinoLogger = ReturnType<typeof Pino>;

/**
 * @category model
 * @since 0.0.1
 */
export type LogLevel = Pino.Level;

/**
 * @category model
 * @since 0.0.1
 */
interface LogFn {
  (obj: unknown, msg?: string, ...args: any[]): IO<void>;
  (msg: string, ...args: any[]): IO<void>;
}

/**
 * @category model
 * @since 0.0.1
 */
export type Logger = ReturnType<typeof Pino>;

// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------

/**
 * @category utils
 * @since 0.0.1
 */
export function fatal(logger: Logger): LogFn {
  return (...args: any): IO<void> =>
    () => {
      logger.fatal.apply(logger, args);
    };
}

/**
 * @category utils
 * @since 0.0.1
 */
export function error(logger: Logger): LogFn {
  return (...args: any): IO<void> =>
    () => {
      logger.error.apply(logger, args);
    };
}

/**
 * @category utils
 * @since 0.0.1
 */
export function warn(logger: Logger): LogFn {
  return (...args: any): IO<void> =>
    () => {
      logger.warn.apply(logger, args);
    };
}

/**
 * @category utils
 * @since 0.0.1
 */
export function info(logger: Logger): LogFn {
  return (...args: any): IO<void> =>
    () => {
      logger.info.apply(logger, args);
    };
}

/**
 * @category utils
 * @since 0.0.1
 */
export function debug(logger: Logger): LogFn {
  return (...args: any): IO<void> =>
    () => {
      logger.debug.apply(logger, args);
    };
}

/**
 * @category utils
 * @since 0.0.1
 */
export function trace(logger: Logger): LogFn {
  return (...args: any): IO<void> =>
    () => {
      logger.trace.apply(logger, args);
    };
}
