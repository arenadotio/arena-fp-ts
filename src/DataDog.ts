/**
 * @since 0.0.1
 */

import * as DI from './DynamicImport';
import * as TO from 'fp-ts/lib/TaskOption';
import { IO } from 'fp-ts/lib/IO';
import { pipe } from 'fp-ts/lib/function';
import { TaskOption } from 'fp-ts/lib/TaskOption';

// -------------------------------------------------------------------------------------
// model
// -------------------------------------------------------------------------------------

/**
 * @category model
 * @since 0.0.1
 */
export interface Metric {
    metricName: string;
    value?: number;
    extraTags?: Record<string, string> | string[];
}

/**
 * @category internal
 */
export type HotShots = typeof import('hot-shots');

/**
 * @category internal
 */
export type HotShotsStatsD = InstanceType<HotShots['StatsD']>

/**
 * @category internal
 */
export type DatadogLambdaJs = typeof import('datadog-lambda-js');

// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------

/**
 * @category instances
 * @since 0.0.1
 */
export const HotShots = DI.fromModule<HotShots>('hot-shots');

/**
 * @category instances
 * @since 0.0.1
 */
export const DatadogLambdaJs = DI.fromModule<DatadogLambdaJs>('datadog-lambda-js');

// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------

/**
 * @category internal
 */
const incrementMetricHotShot = (
    client: HotShotsStatsD
): (a: Metric) => IO<void> => (
    metric
) => () => {
    const { metricName, value, extraTags } = metric;
    const fullMetricName = `${metricName.replace('-', '_').replace(' ', '_')}`;
    client.increment(fullMetricName, value || 1, extraTags);
}

/**
 * @category internal
 */
const incrementMetricLambda = (
    datadog: DatadogLambdaJs
): (a: Metric) => IO<void> => (
    metric
) => () => {
    const { metricName, value, extraTags } = metric;
    const fullMetricName = `${metricName.replace('-', '_').replace(' ', '_')}`;

    const tags = typeof extraTags === 'object' ? Object.entries(extraTags).map(([k, v]) => `${k}:${v}`) : extraTags;

    datadog.sendDistributionMetric(fullMetricName, value || 1, ...(tags || []));
}

/**
 * @category utils
 * @since 0.0.1
 */
export const incrementMetric: (a: Metric) => TO.TaskOption<void> = (a) => {
    const lambda: TaskOption<(a: Metric) => IO<void>> = pipe(
        DatadogLambdaJs,
        TO.map(incrementMetricLambda),
    );

    const hotshot: TaskOption<(a: Metric) => IO<void>> = pipe(
        HotShots,
        TO.map((hotShots: HotShots) => new hotShots.StatsD()),
        TO.map(incrementMetricHotShot),
    );

    return pipe(
        lambda,
        TO.alt(() => hotshot),
        TO.flatMapIO(f => f(a)),
    );
}
