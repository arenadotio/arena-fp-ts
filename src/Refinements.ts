/**
 * @since 0.0.11
 */
import { Refinement } from 'fp-ts/lib/Refinement';
import { Option } from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/function';
import { Either } from 'fp-ts/lib/Either';

/**
 * @category refinements
 * @since 0.0.11
 */
export const isTagged: Refinement<unknown, { _tag: string }> = (
  u
): u is { _tag: string } => !!u && typeof u === 'object' && '_tag' in u;

/**
 * @category refinements
 * @since 0.0.11
 */
export const hasTag =
  <A extends [string, ...string[]]>(
    ...values: A
  ): Refinement<unknown, { _tag: A[number] }> =>
  (u: unknown): u is { _tag: A[number] } =>
    isTagged(u) && values.includes(u._tag);

/**
 * @category refinements
 * @since 0.0.11
 */
export const isOption = <A = unknown>(u: unknown): u is Option<A> =>
  pipe(u, hasTag('Some', 'None'));

/**
 * @category refinements
 * @since 0.0.11
 */
export const isEither = <E = unknown, A = unknown>(
  u: unknown
): u is Either<E, A> => pipe(u, hasTag('Left', 'Right'));
