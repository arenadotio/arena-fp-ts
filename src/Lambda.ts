/**
 * ```
 * export interface State {
 *   appName: string;
 *   logger: L.Logger;
 * }
 * ```
 *
 * ```
 * export type Main<A> = SRTE.StateReaderTaskEither<
 *   State,
 *   A,
 *   [Error, State],
 *   void
 * >;
 * ```
 *
 * ```
 * export interface Lambda<A> {
 *   appName?: string;
 *   logger?: L.Logger;
 *   codec: Type<A, any, unknown>;
 *   main: Main<A>;
 * }
 * ````
 *
 * This module is used to transform a `Lambda<A>` into a function that
 * AWSLambda can call.
 *
 * The handler contains a main function that is implemented as a
 * StateReaderTaskEither<State, A, [Error, State], void>. Not that the left type
 * is also a tuple linking an Error with a new state in the same way that the
 * right type usually does. This allows the State to be uploaded at some point
 * during the handler event if it eventually fails.
 *
 * Before the handler is invoked, the input from AWS will be validated using
 * io-ts. If any validation errors are signaled, the handler will never be
 * called and an exception will be sent to Sentry.
 *
 * @example
 * import * as TE from 'fp-ts/lib/TaskEither'
 * import * as t from 'io-ts';
 *
 * import * as L from 'arena-fp-ts/Lambda';
 *
 * const myCoolLambda: L.Lambda<string> = {
 *   codec: t.string,
 *   main: (state) => (_event) => TE.left([new Error("oof I don't do anything"), state])
 * }
 *
 * export const app = L.toAWSLambda(myCoolLambda);
 *
 * @since 0.0.1
 */

import { Type, Decoder } from 'io-ts';
import * as t from 'io-ts';

import { flow, pipe } from 'fp-ts/lib/function';

import * as IO from 'fp-ts/lib/IO';
import * as IOE from 'fp-ts/lib/IOEither';
import * as TE from 'fp-ts/lib/TaskEither';
import * as E from 'fp-ts/lib/Either';
import * as T from 'fp-ts/lib/Task';
import * as SRTE from 'fp-ts/lib/StateReaderTaskEither';

import * as L from './Logger';
import * as V from './validator';
import * as Sentry from './Sentry';
import * as DD from './DataDog';

import { AWSLambda } from '@sentry/serverless';
import * as AWS from 'aws-lambda';
import { formatValidationErrors } from 'io-ts-reporters';

// -------------------------------------------------------------------------------------
// model
// -------------------------------------------------------------------------------------

/**
 * @category model
 * @since 0.0.6
 */
export type AWSHandler<A> = AWS.Handler<{ detail: A }, void>;

/**
 * @deprecated Use State
 * @category model
 * @since 0.0.1
 */
export interface LambdaState<A> {
  appName: string;
  logger: L.Logger;
  event: A;
}

/**
 * @deprecated Use Main<A>
 * @since 0.0.1
 */
export type Handler<A> = (
  state: LambdaState<A>
) => T.Task<readonly [E.Either<Error, void>, LambdaState<A>]>;

/**
 * @category model
 * @since 0.0.6
 */
export interface State {
  appName: string;
  logger: L.Logger;
}

/**
 * @category model
 * @since 0.0.6
 */
export type Main<A> = SRTE.StateReaderTaskEither<
  State,
  A,
  [Error, State],
  void
>;

/**
 * @category model
 * @since 0.0.6
 */
export interface Lambda<A> {
  appName?: string;
  logger?: L.Logger;
  sentryDsn?: string;
  sentryEnvironment?: string;
  sentryRelease?: string;
  codec: Type<A, any, unknown>;
  main: Main<A>;
}

// -------------------------------------------------------------------------------------
// conversions
// -------------------------------------------------------------------------------------

/**
 * @deprecated Use toAWSLambda()
 * @category conversions
 * @since 0.0.8
 */
export function toLambda<A>(
  appName: string,
  codec: Decoder<unknown, A>,
  handler: Handler<A>
): AWSHandler<A> {
  // Create logger instance and reporting functions
  const logger = L.makeLogger(appName);
  const debug = L.debug(logger);
  const info = L.info(logger);
  const error = Sentry.captureException(logger);

  const incrementMetric = (name: string) =>
    pipe(
      DD.incrementMetric({
        metricName: name,
        extraTags: { app_name: appName },
      }),
      T.asUnit
    );

  // Log that we've started running and initialize sentry
  const initialize = pipe(
    debug('Logger created'),
    IO.apSecond(Sentry.init(logger))
  );

  initialize();

  // Tries to create a LambdaState from the input AWS passed in
  const getInitialState = pipe(
    (event: A): LambdaState<A> => ({ appName, logger, event }),
    V.fromCodec(codec),
    (f) => flow(f, TE.fromEither)
  );

  // Task a validate state and executes the handler
  const run: (state: LambdaState<A>) => T.Task<void> = (state) => {
    const task = pipe(
      T.fromIO(info('Running handler')),
      T.flatMap((_) => handler(state)),
      T.flatMap(([result, { logger }]) => {
        if (E.isLeft(result)) {
          const error = Sentry.captureException(logger);
          return T.fromIO(error(result.left));
        } else {
          const info = L.info(logger);
          return T.fromIO(info('Handler finished successfully'));
        }
      })
    );

    return T.asUnit(task);
  };

  // Parse input from AWS and run the handler if valid
  const processEvent = flow(
    getInitialState,
    TE.fold(flow(error, T.fromIO), run)
  );

  return AWSLambda.wrapHandler((input: unknown, _context, _callback) => {
    const event =
      input && typeof input === 'object' && 'detail' in input
        ? input.detail
        : null;

    const program = T.sequenceSeqArray([
      incrementMetric('start'),
      T.fromIO(info({ event }, 'Received event')),
      processEvent(event),
      T.fromIO(info({ event }, 'Finished processing event')),
      incrementMetric('end'),
    ]);

    return T.asUnit(program)();
  });
}

/**
 * @category conversions
 * @since 0.0.6
 */
export function toAWSLambda<A>(lambda: Lambda<A>): AWSHandler<A> {
  const appName =
    lambda.appName || process.env.npm_package_config_appName || 'Unnamed App';
  const logger = lambda.logger || L.makeLogger(appName);
  const initialState: State = {
    appName,
    logger,
  };
  const EventCodec = t.type({
    detail: lambda.codec,
  });

  const debug = L.debug(logger);
  const error = L.error(logger);
  const incrementMetric = (name: string) =>
    pipe(
      DD.incrementMetric({
        metricName: name,
        extraTags: { app_name: appName },
      }),
      T.asUnit
    );

  const validateInput = (input: unknown): TE.TaskEither<[Error, State], A> =>
    pipe(
      debug('Validating Input'),
      IO.apSecond(() => EventCodec.decode(input)),
      IOE.map((event) => event.detail),
      IOE.mapError(formatValidationErrors),
      IO.tap((result) => {
        if (E.isLeft(result)) {
          return error({ errors: result.left }, 'Input Validation Failed');
        } else {
          return debug({ result: result.right }, 'Input Validation Succeeded');
        }
      }),
      IOE.mapError(
        (errors) =>
          [
            new Error(`Validation Failed: ${errors.join('; ')}`),
            initialState,
          ] as [Error, State]
      ),
      TE.fromIOEither
    );

  const initialize = IO.sequenceArray([
    debug('Logging initialized'),
    Sentry.init(logger, {
      dsn: lambda.sentryDsn,
      release: lambda.sentryRelease,
      environment: lambda.sentryEnvironment,
    }),
  ]);
  initialize();

  return AWSLambda.wrapHandler((input: unknown, _context, _callback) => {
    const program = pipe(
      TE.fromTask(incrementMetric('start')),
      TE.apSecond(validateInput(input)),
      TE.flatMap(lambda.main(initialState)),
      T.map(E.toUnion),
      T.flatMapIO(([result, { logger }]) => {
        const info = L.info(logger);
        const error = L.error(logger);

        if (result instanceof Error) {
          return pipe(
            IO.of(result),
            IO.tap((err) => error({ err }, 'Lambda failed!')),
            IO.tap(Sentry.captureException(logger))
          );
        } else {
          return info('Lambda finished successfully');
        }
      }),
      T.apFirst(incrementMetric('end'))
    );

    return program().catch((err) => {
      error(
        { err },
        `Running the lambda resulted in an exception. This should not be able to happen \
because the main function should only communicate errors with the Either monad.`
      );
      Sentry.captureException(logger)(err)();
      incrementMetric('end');
    });
  });
}
