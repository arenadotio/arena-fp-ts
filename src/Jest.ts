/**
 * In addition to the various types and utility functions, this module registers
 * new matchers into Jest's `expect` module along with the matchers provided by
 * `@relmify/jest-fp-ts`. Importing this module at the type of any test file
 * will make these matchers available, but you can also add to your
 * `jest.config.js`:
 *
 * ```
 * module.exports = {
 *   ...
 *   setupFilesAfterEnv: ['arena-fp-ts/Jest'],
 *   ...
 * };
 * ```
 *
 * @since 0.0.5
 */
import { matchers } from '@relmify/jest-fp-ts';

import * as SRTE from 'fp-ts/lib/StateReaderTaskEither';
import * as RTE from 'fp-ts/lib/ReaderTaskEither';
import * as TE from 'fp-ts/lib/TaskEither';
import * as A from 'fp-ts/lib/Array';
import * as O from 'fp-ts/lib/Option';

import * as L from './Logger';

import { Type } from 'io-ts';

import { expect } from '@jest/globals';
import { equals } from '@relmify/jest-fp-ts/dist/predicates/equals';
import { isEitherOrThese } from '@relmify/jest-fp-ts/dist/predicates/isEitherOrThese';
import {
  EitherOrThese,
  fold,
  isRight,
} from '@relmify/jest-fp-ts/dist/eitherOrThese/eitherOrThese';
import { constFalse, flow, pipe } from 'fp-ts/lib/function';
import {
  matcherHint,
  printDiffOrStringify,
  printExpected,
  printReceived,
} from 'jest-matcher-utils';
import { State } from './Lambda';

// -------------------------------------------------------------------------------------
// model
// -------------------------------------------------------------------------------------

/**
 * @category models
 * @since 0.0.5
 */
export type StateReaderTaskEitherPropertyNames<T> = {
  [K in keyof T]: T[K] extends SRTE.StateReaderTaskEither<any, any, any, any>
    ? K
    : never;
}[keyof T] &
  string;

// -------------------------------------------------------------------------------------
// utils
// -------------------------------------------------------------------------------------

/**
 * Spies on a StateReaderTaskEither and mocks the ReaderTaskEither that it
 * returns.
 *
 * ```
 * import { pipe } from 'fp-ts/lib/function';
 * import * as TE from 'fp-ts/lib/TaskEither';
 * import * as E from 'fp-ts/lib/Either';
 * import { mockStateReaderTaskEither } from 'arena-fp-ts/Jest';
 *
 * import * as fakeModule from 'my-fake-module';
 *
 * describe(mockStateReaderTaskEither, () => {
 *     it('can mock on StateReaderTaskEither return values', async () => {
 *         const [spy] = mockStateReaderTaskEither(fakeModule, 'testFunction');
 *         spy.mockReturnValue(TE.of([10, 'foo']));
 *
 *         const program = pipe(
 *             1,
 *             fakeModule.testFunction('some state'),
 *         );
 *         const result = await program();
 *
 *         expect(E.isRight(result) && result.right).toEqual([10, 'foo']);
 *     });
 *
 *     it('can spy on reader arguments', async () => {
 *         const [spy] = mockStateReaderTaskEither(fakeModule, 'testFunction');
 *
 *         const program = pipe(
 *             1,
 *             fakeModule.testFunction('some state'),
 *         );
 *         await program();
 *
 *         expect(spy).toHaveBeenCalledWith(1);
 *     });
 *
 *     it('can spy on state arguments', async () => {
 *         const [_, spy] = mockStateReaderTaskEither(fakeModule, 'testFunction');
 *
 *         const program = pipe(
 *             1,
 *             fakeModule.testFunction('some state'),
 *         );
 *         await program();
 *
 *         expect(spy).toHaveBeenCalledWith('some state');
 *     });
 * });
 * ```
 *
 * @category utils
 * @since 0.0.5
 */
export function mockStateReaderTaskEither<
  S,
  R,
  E,
  A,
  T extends Record<string, unknown>,
  M extends StateReaderTaskEitherPropertyNames<Required<T>>
>(
  object: T,
  property: M
): Required<T>[M] extends SRTE.StateReaderTaskEither<any, any, any, any>
  ? [
      jest.SpyInstance<TE.TaskEither<E, [A, S]>, [S]>,
      jest.SpyInstance<RTE.ReaderTaskEither<R, E, [A, S]>, [S]>
    ]
  : never {
  const fn = jest.fn(() => TE.of(undefined));

  const spy = jest.spyOn(object, property as any);

  spy.mockReturnValue(fn as any);

  return [fn, spy] as any;
}

/**
 * @category utils
 * @since 0.0.6
 */
export function mockLambda<A>(
  codec: Type<A, any, unknown>,
  returnValue: (state: State) => TE.TaskEither<[Error, State], [void, State]>
): [
  {
    codec: Type<A, any, unknown>;
    main: jest.Mock<
      RTE.ReaderTaskEither<A, [Error, State], [void, State]>,
      [State]
    >;
  },
  jest.Mock<TE.TaskEither<[Error, State], [void, State]>, [A]>
] {
  const appName = 'mock';
  const logger = L.makeLogger(appName);
  const state: State = {
    appName,
    logger,
  };
  const ret = returnValue(state);
  const rte = jest.fn((_: A) => ret);
  const main = jest.fn((_: State) => rte);

  return [
    {
      codec,
      main,
    },
    rte,
  ] as any;
}

// -------------------------------------------------------------------------------------
// Jest matchers
// -------------------------------------------------------------------------------------

/** @internal */
const applyPredicateRightTuple =
  (accessor: (as: unknown[]) => O.Option<unknown>) =>
  (predicate: (rightResultValue: unknown) => boolean) =>
  (received: EitherOrThese<unknown, unknown>): boolean =>
    pipe(
      received,
      fold(
        constFalse,
        flow(
          O.fromPredicate(Array.isArray),
          O.flatMap(accessor),
          O.filter(predicate),
          O.isSome
        ),
        constFalse
      )
    );

/** @internal */
const applyEqualsRightTupleMatcher =
  (matcherName: string, accessor: (as: unknown[]) => O.Option<unknown>) =>
  (received: unknown, expected: unknown): any => {
    const applyPredicate = pipe(
      equals(expected),
      applyPredicateRightTuple(accessor)
    );

    if (!isEitherOrThese(received)) {
      return {
        pass: false,
        message: () =>
          [
            matcherHint(matcherName, 'received', 'expected'),
            '\n',
            'Received value is not an Either or These.',
            `Received: ${printReceived(received)}`,
          ].join('\n'),
      };
    } else if (!isRight(received)) {
      return {
        pass: false,
        message: () =>
          [
            matcherHint(matcherName, 'received', 'expected'),
            '\n',
            'Received value is not Right.',
            `Received: ${printReceived(received)}`,
          ].join('\n'),
      };
    } else if (!Array.isArray(received.right) || received.right.length !== 2) {
      return {
        pass: false,
        message: () =>
          [
            matcherHint(matcherName, 'received', 'expected'),
            '\n',
            'Received value is not a tuple: [result, state].',
            `Received: ${printReceived(received.right)}`,
          ].join('\n'),
      };
    } else if (!applyPredicate(received)) {
      return {
        pass: false,
        message: () =>
          [
            matcherHint(matcherName, 'received', 'expected'),
            '\n',
            printDiffOrStringify(
              expected,
              O.toNullable(accessor(received.right as any)),
              'Expected',
              'Received.right',
              true
            ),
          ].join('\n'),
      };
    } else {
      return {
        pass: true,
        message: () =>
          [
            matcherHint('.not.' + matcherName, 'received', 'expected'),
            '\n',
            `Expected: not ${printExpected(expected)}`,
          ].join('\n'),
      };
    }
  };

/**
 * Checks that the received value is a `Right` `Either` or `These` and that its
 * value is a tuple `[result, state]`. It finally checks that the `result` is
 * equal to the expected value using Jest's `toEqual()`.
 *
 * This is useful when testing the `State` monad or its derivatives.
 *
 * ```
 * describe('toHaveRightValueWithResult', () => {
 *   it('matches against a right result value', () => {
 *     expect(E.right([{ foo: 'bar' }, 'some state'])).toHaveRightValueWithResult({
 *       foo: 'bar',
 *     });
 *   });
 * });
 * ```
 *
 * @category matchers
 * @since 0.0.5
 */
export const toHaveRightValueWithResult = applyEqualsRightTupleMatcher(
  'toHaveRightValueWithResult',
  A.head
);

/**
 * Checks that the received value is a `Right` `Either` or `These` and that its
 * value is a tuple `[result, state]`. It finally checks that the `state` is
 * equal to the expected value using Jest's `toEqual()`.
 *
 * This is useful when testing the `State` monad or its derivatives.
 *
 * ```
 * describe('toHaveRightValueWithState', () => {
 *   it('matches against a right result value', () => {
 *     expect(E.right([{ foo: 'bar' }, 'some state'])).toHaveRightValueWithState(
 *       'some state'
 *     );
 *   });
 * });
 * ```
 *
 * @category matchers
 * @since 0.0.5
 */
export const toHaveRightValueWithState = applyEqualsRightTupleMatcher(
  'toHaveRightValueWithState',
  A.lookup(1)
);

/** @internal */
type Matchers<T = typeof matchers> = {
  [K in keyof T]: T[K] extends (...args: [any, ...infer R]) => any
    ? (...args: R) => any
    : never;
};

/** @internal */
type CustomMatchers<R = unknown> = Matchers & {
  /**
   * Use `.toHaveRightValueWithResult` to check if a value is an `Either` or
   * `These`, that it is `Right`, its value is a tuple `[result, state]`, and
   * that the `result` is equal to an expected value.
   *
   * Equality is checked with Jest's `.toEqual()` comparison.
   */
  toHaveRightValueWithResult(expected: unknown): R;

  /**
   * Use `.toHaveRightValueWithResult` to check if a value is an `Either` or
   * `These`, that it is `Right`, its value is a tuple `[result, state]`, and
   * that the `state` is equal to an expected value.
   *
   * Equality is checked with Jest's `.toEqual()` comparison.
   */
  toHaveRightValueWithState(expected: unknown): R;
};

expect.extend({
  toHaveRightValueWithResult,
  toHaveRightValueWithState,
  ...matchers,
});

declare module 'expect' {
  /* eslint-disable-next-line @typescript-eslint/no-empty-interface */
  interface Matchers<R> extends CustomMatchers<R> {}
}
