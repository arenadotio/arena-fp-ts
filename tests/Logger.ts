import * as L from '../src/Logger';
import * as S from 'fp-ts/lib/State';
import * as A from 'fp-ts/lib/ReadonlyArray';
import { IO, Apply, map, apSecond } from 'fp-ts/lib/IO';
import { sequenceT } from 'fp-ts/lib/Apply';
import Pino from 'pino';
import { pipe } from 'fp-ts/lib/function';

describe('Logging', () => {
  const message = 'This is a message %o';
  const mergeObject = { foo: 'bar' };
  const interpolationValue = { bar: 'foo' };

  const spy = jest.fn();

  afterEach(() => {
    spy.mockClear();
  });

  // Logger invocation
  it.each(['fatal', 'error', 'warn', 'info', 'debug', 'trace'] as L.LogLevel[])(
    'Log level %s',
    (level) => {
      const logger = Pino({
        level: 'trace',
      });

      logger[level] = spy;

      const log = L[level](logger);

      const argsList = [
        [message],
        [message, interpolationValue],
        [mergeObject],
        [mergeObject, message],
        [mergeObject, message, interpolationValue],
      ] as const;

      for (const args of argsList) {
        const io = log.apply(null, args as any);
        io();
        expect(spy).toHaveBeenLastCalledWith(...args);
      }
    }
  );

  // Logger composition
  it('Can be composed', () => {
    const inc =
      (x: number): S.State<L.Logger, IO<number>> =>
      (logger) => {
        const log = L.info(logger);
        return [
          pipe(
            log('called with %d', x),
            apSecond(() => x + 1)
          ),
          logger.child({ ['inc' + x]: 'called' }),
        ];
      };

    const steps = [inc(1), inc(2), inc(3)];

    const messages: any[] = [];
    const initialLogger = Pino({
      hooks: {
        logMethod(args, method, _level) {
          messages.push(args);
          return method.apply(this, args);
        },
      },
    });

    const [program, finalLogger] = pipe(
      initialLogger,
      S.sequenceArray(steps),
      ([ios, finalLogger]) => {
        const result = pipe(
          /* eslint-disable @typescript-eslint/no-non-null-assertion */
          [ios[0]!, ...ios.slice(1)] as const,
          (args) => sequenceT(Apply)(...args),
          map(A.last)
        );

        return [result, finalLogger] as const;
      }
    );

    finalLogger.info('one last message');
    finalLogger.flush();

    expect(program()).toMatchObject({ value: 4 });
    expect(messages).toMatchSnapshot();
    expect(finalLogger.bindings()).toMatchSnapshot();
  });
});
