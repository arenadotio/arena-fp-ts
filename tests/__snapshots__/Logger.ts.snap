// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logging Can be composed 1`] = `
[
  [
    "one last message",
  ],
  [
    "called with %d",
    1,
  ],
  [
    "called with %d",
    2,
  ],
  [
    "called with %d",
    3,
  ],
]
`;

exports[`Logging Can be composed 2`] = `
{
  "inc1": "called",
  "inc2": "called",
  "inc3": "called",
}
`;
